File: coltsteele/local_from_work_laptop/multiple_pointers/multiple_pointers_isSubsequence.js
// Udemy
// JavaScript Algorithms and Data Structures Masterclass
// Multiple Pointers - isSubsequence
// Write a function called isSubsequence which takes in two strings and checks whether the characters in the first string form a subsequence of the characters in the second string. In other words, the function should check whether the characters in the first string appear somewhere in the second string, without their order changing.

// Examples:

// isSubsequence('hello', 'hello world'); // true
// isSubsequence('sing', 'sting'); // true
// isSubsequence('abc', 'abracadabra'); // true
// isSubsequence('abc', 'acb'); // false (order matters)
// Your solution MUST have AT LEAST the following complexities:

// Time Complexity - O(N + M)

// Space Complexity - O(1)

function isSubsequence(word1, word2) {
    let count = 0
    for (let index1 = 0, index2 = 0; index1 < word1.length && index2 < word2.length; index1) {
        const letter1 = word1[index1];
        let letter2 = word2[index2];
        if(letter1 == letter2) {
            count++;
            index1++;
            index2++;
        }
        if(letter1 != letter2){
            index2++;
        }
    }
    if (count == word1.length) return true;
    return false;
}

//Colt Steele Solution
function isSubsequence(str1, str2) {
    var i = 0;
    var j = 0;
    if (!str1) return true;
    while (j < str2.length) {
      if (str2[j] === str1[i]) i++;
      if (i === str1.length) return true;
      j++;
    }
    return false;
  }

  // This solution is similar to mine but smarter. 
  // It uses the count as the index. smart smart

File: coltsteele/local_from_work_laptop/multiple_pointers/multiple_pointers_is_subsequence.rb
# Udemy
# JavaScript Algorithms and Data Structures Masterclass
# Multiple Pointers - is_subsequence
# Write a function called is_subsequence which takes in two strings and checks whether the characters in the first string form a subsequence of the characters in the second string. In other words, the function should check whether the characters in the first string appear somewhere in the second string, without their order changing.

# Examples:

# is_subsequence('hello', 'hello world'); # true
# is_subsequence('sing', 'sting'); # true
# is_subsequence('abc', 'abracadabra'); # true
# is_subsequence('abc', 'acb'); # false (order matters)
# Your solution MUST have AT LEAST the following complexities:

# Time Complexity - O(N + M)

# Space Complexity - O(1)

def is_subsequence(word1, word2) 
    count, index2 = 0, 0
    while index2 < word2.length && count < word1.length
        letter1, letter2 = word1[count], word2[index2]
        if (letter1 == letter2)
            count += 1; index2 += 1 #Can't do single line if because the ; breaks the line
        end
        index2 += 1 if(letter1 != letter2)
    end
    return true if (count == word1.length)
    false
end

def is_subsequence(word1, word2) 
    count = 0
    for i in 0...word2.length
        puts "#{word1[count]} - #{word2[i]}"
        count += 1 if word1[count] == word2[i]
        return true if count == word1.length
    end
    false
end

# for i in 0..5
#     if i == 2
#         i = 4
#         redo
#     end
#     puts "Value of local variable is #{i}"
# end

# Value of local variable is 0
# Value of local variable is 1
# Value of local variable is 4
# Value of local variable is 3
# Value of local variable is 4
# Value of local variable is 5

def test
    puts("#{is_subsequence('hello', 'hello world') == true}")
    puts("#{is_subsequence('sing', 'sting') == true}")
    puts("#{is_subsequence('abc', 'abracadabra') == true}")
    puts("#{is_subsequence('abc', 'acb') == false }")
end 

#Colt Steele Solution
def is_subsequence(str1, str2)
    i = 0
    j = 0
    if (!str1) return true;
    while (j < str2.length) do
      if (str2[j] == str1[i]) i += 1
      if (i == str1.length) return true
      j += 1
    end
    return false
end

  # This solution is similar to mine but smarter. 
  # It uses the count as the index. smart smart